### DISCUSSION TITLE
BusinessRule primary property as IPropertyInfo or PropertyInfo<>

### DISCUSSION THREAD

**Discussion by joshhanson314** [MarimerLLC/csla#3693] [Questions] [answered]
In every sample I see, properties are sent to the business rules as IPropertyInfo. Is there a reason to NOT create business rules with a constructor parameter of PropertyInfo<> instead? For example, I have a rule that expects the primary property to be a string. I can enforce this at the caller by using `BusinessRules.Add(new MyRule(MyStringProperty));` where MyStringProperty is `PropertyInfo<string>` and not have to worry about bad casting (improper assignment of the rule property) in the execute.

**Comment by hurcane** [comment, marked as answer]
You are free to design the constructors on your rule classes as best fits your needs. PrimaryProperty of the rule class is type IPropertyInfo. As long as the type used on the constructor parameter implements IPropertyInfo, you can assign it to PrimaryProperty.

Within the Execute method, references to PrimaryProperty would still need to be cast. You could keep a secondary private variable of the PropertyInfo<string> type to avoid using PrimaryProperty.

**Reply by rockfordlhotka** [reply]
I suppose there could be a `BusinessRule<T>` base class where `T` is the type of the primary property. Would that be helpful?

**Reply by joshhanson314** [reply]
@hurcane I have done that many times over already. My hope was to see if there was a specific reason to NOT do that.

@rockfordlhotka That would be pretty cool. To me, this definitively says "there is no reason to not type them." Would you like me to submit a PR? (Perhaps this is fair trade to ward off the ban hammer from #3689 😇)

**Reply by rockfordlhotka** [reply]
Please write up the proposal as an Issue so we can agree on the solution, then yes, a PR would be nice!
